aiohttp/streams.py

```diff
@@ -320,16 +347,34 @@ def readany(self):
 
     @asyncio.coroutine
     def readchunk(self):
+        """Returns a tuple of (data, end_of_http_chunk). When chunked transfer
+        encoding is used, end_of_http_chunk is a boolean indicating if the end
+        of the data corresponds to the end of a HTTP chunk , otherwise it is
+        always False.
+        """
         if self._exception is not None:
             raise self._exception
 
         if not self._buffer and not self._eof:
+            # end of http chunk without available data
+            if self._http_chunk_splits and \
+                    self._cursor == self._http_chunk_splits[0]:
+                self._http_chunk_splits = self._http_chunk_splits[1:]
+                return (b"", True)
             yield from self._wait('readchunk')
 
-        if self._buffer:
-            return self._read_nowait_chunk(-1)
+        if not self._buffer:
+            # end of file
+            return (b"", False)
+        elif self._http_chunk_splits is not None:
+            if self._http_chunk_splits:
+                pos = self._http_chunk_splits[0]
+                if pos > self._cursor:
```

Comment:
FTR this condition was here because I was afraid of what would happen if some other calls than `readchunk` were made before. 
So I did a small modification to the condition here and added tests for those situations, covering cases where `stream.unread_data` and `stream.read` are used before `stream.readchunk`.